doc/configuration.md
Configuration Guide
===================

Lactoserv is configured with a JavaScript source file (notably, as opposed to a
JSON file). This tactic is meant to remove the need for "halfhearted
programming" facilities baked into a system configuration parser itself. A
configuration file is expected to be a module (`.mjs` or `.cjs`) which has a
single `default` export consisting of a JavaScript object of the ultimate
configuration. Very skeletally (and reductively):

```js
const config = {
  services: [
    // ... configuration ...
  ],
  hosts: [
    // ... configuration ...
  ],
  endpoints: [
    // ... configuration ...
  ],
  applications: [
    // ... configuration ...
  ]
};

export default config;
```

**Note:** Please refer to the
[example configuration file](../etc/example-setup/config/config.mjs)
for reference: First, while this guide is intended to be accurate, the example
configuration is actually tested regularly. Second, the example uses some of the
tactics which are mentioned here, so you can see them "in action."

## `import`s

All core Node libraries are available for `import` in a configuration file,
but they _must_ be imported using the `node:` prefix form (not just plain
names), e.g.:

```js
import { readFile } from 'node:fs/promises';
```

In addition, all of Lactoserv's framework classes are available for import,
using the module namespace prefix `@lactoserv/`, e.g.:

```js
import { Moment } from '@lactoserv/data-values';
```

And in order to use any of the built-in applications or services, you will need
to import them from `@lactoserv/built-ins`.

## Component instantiation

Lactoserv configuration boils down to defining a tree of "component" objects.
These objects can be in one of two forms:

* A configuration object &mdash; a plain object &mdash; with all the required
  bindings of the component in question, along with an extra property `class`
  which indicates the class of the object (as a class object / constructor
  function, _not_ just a string name).
* A directly instantiated object, using the usual `new ClassName(...)` syntax,
  passing it a plain object of its configuration bindings.

In the rest of this guide, we use the plain object form, which arguably feels
more natural for a configuration file per se. When using Lactoserv as a
framework, though, the direct instantiation form is probably to be preferred.

```js
// Plain object form.
const config = {
  applications: [
    {
      class: StaticFiles,
      // ... more ....
    },
    // ... more ...
  ],
  // ... more ...
};
```

```js
// Direct instantiation form.
const warehouse = new Warehouse({
  applications: [
    new StaticFiles({
      // ... more ...
    }),
    // ... more ...
  ],
  // ... more ...
});
```

## Configuration object bindings

The following are the bindings that are expected at the top level of the
exported configuration object. In each case, the binding is described as a
"list" and is typically a JavaScript array. However, in cases where only one
element is needed, a plain object may be bound directly instead of being a
one-element array.

### `hosts`

`hosts` is a list of hostname bindings. These map possibly-wildcarded hostnames
to certificate-key pairs to use to authenticate an endpoint as those hosts. Each
entry has the following bindings:

* `hostnames` &mdash; A list of one or more hostnames to recognize; this is
   required. Hostnames are allowed to start with a `*` to indicate a wildcard of
   _any number of subdomains, including zero._ Note that this is unlike how
   wildcards work in the underlying certificates, where a `*` denotes exactly
   one subdomain. And, to be clear, the hostname `*` will match _any_ hostname
   at all, with any number of subdomains.
* `certificate` &mdash; PEM format string containing the certificate to use for
  this entry. This is required if `selfSigned` is absent or `false`.
* `privateKey` &mdash; PEM format string containing the private key to use for
  this entry. This is required if `selfSigned` is absent or `false`.
* `selfSigned` &mdash; Optional boolean, which, if `true`, causes the system to
  generate a self-signed certificate for this entry. This is mostly useful in
  testing scenarios, and more specifically when running a server on your
  development machine, e.g. and commonly responding on `localhost`.

```js
const hosts = [
  {
    hostnames:  ['localhost', '*'],
    selfSigned: true
  },
  {
    hostnames:   ['*.example.com'],
    certificate: '-----BEGIN CERTIFICATE-----...',
    privateKey:  '-----BEGIN PRIVATE KEY-----...'
  },
  // ... more ...
];
```

This section is only required if at least one endpoint is to respond to
host-authenticated protocols (which is nearly always, at least in standalone
uses).

**Note:** If you want to keep the text of the keys and certificates out of the
main configuration file, then a reasonablhy easy tactic is to use the standard
Node `fs` package to read the contents of files named in the configuration.

### `services`

`services` is a list of system services to be used, with each element naming and
configuring one of them. A system service is simply an encapsulated bit of
functionality that gets hooked up to the system in general or to some other more
specific part of the system (typically, to one or more network endpoints).

There are two required bindings for each system service, its `name` and its
`class` (type). Beyond that, the configuration depends on the `class`. See below
for a list of all built-in system services. The `name` is used both when logging
activity (to the system log) and when hooking services up.

```js
import { ServiceClass } from '@lactoserv/built-ins';

const services = [
  {
    name:       'someService',
    class:      ServiceClass,
    // ... class-specific configuration ...
  },
  // ... more ...
```

This section is only required if there are any services being defined at all.

### `applications`

`applications` is a list of applications to be used, with each element naming
and configuring one of them. An application is an encapsulated bit of
functionality which specifically knows how to respond to external (HTTP-ish)
requests.

As with services, there are two required bindings for each application, its
`name` and its `class` (type). Beyond that, the configuration depends on the
`class`. See below for a list of all built-in applications. The `name` is used
both when logging activity (to the system log) and when hooking applications up
to endpoints.

```js
import { ApplicationClass } from '@lactoserv/built-ins';

const applications = [
  {
    name:       'someApplication',
    class:      ApplicationClass,
    // ... class-specific configuration ...
  },
  // ... more ...
```

### `endpoints`

`endpoints` is a list of network endpoints to listen on, with each element
naming and configuring one of them. Each element has the following bindings:

* `name` &mdash; The name of the endpoint. This is just used for logging and
  related informational purposes.
* `hostnames` &mdash; A list of one or more hostnames to recognize, each name
  in the same form as accepted in the `hosts` section of the configuration.
  Defaults to `['*']`, which should suffice in most cases.
* `interface` &mdash; The network interface to listen on. This is a string which
  can take one of two forms:
  * `<address>:<port>` &mdash; Specifies a normal network-attached interface.
    `<address>` is a DNS name, an IPv4 address, a _bracketed_ IPv6 address, or
    the wildcard value `*`. `<port>` is a non-zero (decimal) port number.
    **Note:** It is invalid to use the IP-version-specific "any" identifiers
    `::` or `0.0.0.0` (or similar).
  * `/dev/fd/<fd-num>` &mdash; Specifies a file descriptor which is expected to
    already correspond to an open server socket (e.g. set up by `systemd`).
    `<fd-num>` is an arbitrary (decimal) number in the range of valid file
    descriptors.
* `protocol` &mdash; The protocol to speak. This can be any of `http`, `https`,
  or `http2`. `http2` includes fallback to `https`.
* `services` &mdash; An object which binds roles to system services by name.
  This binding is optional, and if present all roles are optional. The following
  roles are recognized:
  * `rateLimiter` &mdash; A request/data rate limiter.
  * `requestLogger` &mdash; A request logger.
* `application`: &mdash; The name of the application which this endpoint should
  send requests to. **Note:** In order to serve multiple leaf applications, the
  one named here will have to be a routing application of some sort (such as
  [`PathRouter`](#pathrouter), for example).

```js
const endpoints = [
  {
    name: 'someEndpoint',
    endpoint: {
      hostnames: ['*'],
      interface: '*:8443',
      protocol:  'http2'
    },
    services: {
      rateLimiter:   'limiter',
      requestLogger: 'requests'
    },
    application: 'mySite'
  },
```

**A note about application mounts:** When finding an application to dispatch a
request to, the system will pick the most-specific matching hostname from the
mounts, and then within that hostname will pick the most-specific matching path.
It then asks the so-identified application to handle the request. If the
application chooses not to handle it, then the system will pick the next most
specific path, and so on, until there are no more options within the selected
hostname. (It will not fall back to less specific hostnames.)

## Built-in Applications

### Common Application Configuration

Most built-in applications implement a set of common configuration options,
which are generally about filtering out or automatically responding to certain
kinds of requests. Exceptions to the use of these configurations are noted in
the documentation of applications, as appropriate. Here are the options:

* `acceptMethods` &mdash; Array of strings indicating which request methods to
  accept. The array can include any of `connect`, `delete`, `get`, `head`,
  `options`, `patch`, `post`, `put`, and/or `trace`. Defaults to the entire set.
* `maxPathDepth` &mdash; Number indicating the maximum (inclusive) allowed
  length _in path components_ of a dispatched request path not including the
  mount point, and not including the empty path component at the end of a
  directory path. `null` indicates "no limit." Defaults to `null`.
* `maxPathLength` &mdash; Number indicating the maximum (inclusive) allowed
  length of the dispatched request path _in octets_. `null` indicates "no
  limit." `0` indicates that no additional path is ever accepted (including even
  a directory slash). Defaults to `null`.
* `maxQueryLength` &mdash; Number indicating the maximum (inclusive) allowed
  length of the query (a/k/a "search") portion of a request URI _in octets_,
  including the leading question mark (`?`). `null` indicates "no limit." `0`
  indicates that queries are not ever accepted. Defaults to `null`.
* `redirectDirectories` &mdash; Boolean indicating whether or not directory
  paths (those ending with an empty path component) should be automatically
  redirected to the file (non-directory) version of the path. Defaults to
  `false`.
* `redirectFiles` &mdash; Boolean indicating whether or not file paths (those
  not ending with an empty path component) should be automatically redirected to
  the directory version of the path. Defaults to `false`.

With regards to the `redirect*` options:
* It is an error to specify both as `true`.
* The redirection (or not) is entirely based on the form of the path, not on any
  file contents (for example). Notably, [`StaticFiles`](#staticfiles) does
  _content_-driven redirection, which is different than what is done here.

With regards to the other options, when a request is filtered out, the result is
that the application simply _doesn't handle_ the request, meaning that the
request will get re-dispatched to the next application in its routing chain (if
any).

```js
import { BaseApplication } from '@lactoserv/sys-framework';

class MyApplication extends BaseApplication {
  // ... more ...
}

const applications = [
  {
    name:           'myCustomApp',
    class:          MyApplication,
    acceptMethods:  ['delete', 'put'],
    maxQueryLength: 0
  }
];
```

### `HostRouter`

An application which can route requests to another application, based on the
`host` (or equivalent) header in the requests. In addition to the
[common application configuration](#common-application-configuration) options,
it accepts the following bindings:

* `hosts` &mdash; A plain object with possibly-wildcarded hostnames as keys, and
  the _names_ of other applications as values. A wildcard only covers the prefix
  of a hostname and cannot be used for hostnames identified by numeric IP
  address.

**Note:** Unlike `PathRouter`, this application does not do fallback to
less-and-less specific routes; it just finds (at most) one to route to.

```js
import { HostRouter } from '@lactoserv/built-ins';

const applications = [
  {
    name:  'myHosts',
    class: HostRouter,
    hosts: {
      '*':             'myCatchAllApp',
      '127.0.0.1':     'myLocalhostApp',
      'localhost':     'myLocalhostApp',
      '*.example.com': 'myExampleApp'
    }
  },
  {
    name: 'myCatchallApp',
    // ... more ...
  },
  {
    name: 'myExampleApp',
    // ... more ...
  },
  {
    name: 'myLocalhostApp',
    // ... more ...
  }
];
```

### `PathRouter`

An application which can route requests to another application, based on the
path of the requests. In addition to the
[common application configuration](#common-application-configuration) options,
it accepts the following bindings:

* `paths` &mdash; A plain object with possibly-wildcarded paths as keys, and
  the _names_ of other applications as values. A wildcard only covers the suffix
  of a path; it cannot be used for prefixes or infixes.

The routing works by starting with the most specific match to the path of an
incoming request. If that app does not try to handle the request &mdash;
note that it counts as a "try" to end up `throw`ing out of the handler &mdash;
the next most specific match is asked, and so on, until there are no path
matches left.

```js
import { PathRouter } from '@lactoserv/built-ins';

const applications = [
  {
    name:  'myPaths',
    class: PathRouter,
    paths: {
      '/*':         'myCatchAllApp',
      '/file':      'myFileApp',
      '/dir/':      'myDirApp',
      '/general/*': 'myGeneralApp',
    }
  },
  {
    name: 'myCatchallApp',
    // ... more ...
  },
  {
    name: 'myFileApp',
    // ... more ...
  },
  {
    name: 'myDirApp',
    // ... more ...
  },
  {
    name: 'myGeneralApp',
    // ... more ...
  }
];
```

### `Redirector`

An application which responds to all requests with an HTTP "redirect" response.
In addition to the
[common application configuration](#common-application-configuration)
options, it accepts the following bindings:

* `acceptMethods` &mdash; Common configuration option, but in this case the
  default is `['delete', 'get', 'head', 'patch', 'post', 'put']`.
* `statusCode` &mdash; Optional HTTP status code to respond with. If not
  specified, it defaults to `301` ("Moved Permanently").
* `target` &mdash; The base URL to redirect to. This is prepended to the partial
  path of each request to form the final redirection URL.
* `cacheControl` &mdash; `cache-control` header definition. If present and not
  `false`, every cacheable response comes with the specified header.

```js
import { Redirector } from '@lactoserv/built-ins';

const applications = [
  {
    name:          'myRedirector',
    class:         Redirector,
    statusCode:    308,
    target:        'https://example.com/boop/',
    cacheControl:  { public: true, maxAge: '5 min' },
    acceptMethods: ['head', 'get']
  }
];
```

### `SerialRouter`

An application which routes requests to one of a list of applications, which are
tried in order. (This is the default / built-in routing strategy of the most
common Node web application frameworks.) In addition to the
[common application configuration](#common-application-configuration) options,
it accepts the following bindings:

* `applications` &mdash; An array listing the _names_ of other applications as
  values.

The routing works by starting with the first element of `applications`, asking
it to handle an incoming request. If that app does not try to handle the request
&mdash; note that it counts as a "try" to end up `throw`ing out of the handler
&mdash; the next application in the list is asked, and so on, until the final
one has been tried.

```js
import { SerialRouter } from '@lactoserv/built-ins';

const applications = [
  {
    name:         'mySeries',
    class:        SerialRouter,
    applications: ['firstApp', 'secondApp']
  },
  {
    name: 'firstApp',
    // ... more ...
  },
  {
    name: 'secondApp',
    // ... more ...
  }
];
```

### `SimpleResponse`

An application which only ever sends one particular response. It's approximately
like `StaticFiles`, except just one file. In addition to the
[common application configuration](#common-application-configuration) options,
it accepts the following configuration bindings:

* `acceptMethods` &mdash; Common configuration option, but in this case the
  default is `['get', 'head']`.
* `body` &mdash; Optional body contents to respond with. If specified, this must
  be either a string or a Node `Buffer` object.
* `contentType` &mdash; Content type to report. This can be either a MIME type
  per se (e.g. `text/plain`) or a commonly-understood extension (with leading
  dot, e.g., `.txt` or `.html`). This must be specified if `body` is. If this is
  not specified but `filePath` is, then the type is inferred from the extension
  on the path. If neither `body` nor `filePath` is specified (that is, for an
  empty body), then this must not be specified either. In the MIME type form,
  an explicit `charset` is honored (e.g. `text/plain; charset=iso-8859-1`), but
  if not specified then `utf-8` is assumed for text types.
* `cacheControl` &mdash; `cache-control` header definition. If present and not
  `false`, every cacheable response comes with the specified header.
* `etag` &mdash; ETag-generating options. If present and not `false`, the
  response comes with an `ETag` header. See "ETag Configuration" below for
  details.
* `filePath` &mdash; Optional absolute filesystem path to the file to respond
  with.
* `statusCode` &mdash; Optional fixed status code to report. If present and not
  `null`, this is the numeric status code that will be used for all responses.
  Setting this also prevents range requests and not-modified responses from
  being generated.

It is valid to specify neither `body` nor `filePath`; this indicates that the
application should only ever produce no-content (status `204`) responses. It is
_not_ valid to specify _both_ `body` and `filePath`.

**Note:** Passing `body` as an empty string or `Buffer` is treated as
zero-length but contentful, e.g. a regular successful response will be status
`200` with `Content-Length: 0`. Likewise, this is how an empty file pointed at
by a `filePath` behaves.

```js
import { SimpleResponse } from '@lactoserv/built-ins';

const applications = [
  {
    name:                'literal',
    class:               SimpleResponse,
    contentType:         'text/plain',
    body:                'Hello!\n',
    cacheControl:        { public: true, maxAge: '1_day' },
    maxPathDepth:        0,
    redirectDirectories: true,
  },
  {
    name:         'fromFile',
    class:        'SimpleResponse',
    filePath:     '/etc/site/notFoundMessage.txt',
    cacheControl: 'immutable; max-age=1000',
    statusCode:   404
  },
  {
    name:  'empty',
    class: 'SimpleResponse'
  }
];
```

There are a few notable things which _aren't_ configurable for this application.
These may be configurable in a future version, if there is sufficient and
reasonable demand:

* Content types:
  * The mapping from extensions to MIME types is not configurable.
* Caching:
  * The `Last-Modified` response header is always sent when given a `filePath`,
    and not sent when given a `body`.
  * Unless `statusCode` is set in the configuration, conditional request headers
    are honored.
* Ranges:
  * Unless `statusCode` is set in the configuration:
    * The `Accept-Ranges` response header is always sent.
    * Range request headers are honored, including conditional range requests.

### `StaticFiles`

An application which serves static files from a local directory. In addition to
most of the
[common application configuration](#common-application-configuration) options
(all but the `redirect*` options, which could cause chaos in this case), it
accepts the following configuration bindings:

* `acceptMethods` &mdash; Common configuration option, but in this case the
  default is `['get', 'head']`.
* `etag` &mdash; ETag-generating options. If present and not `false`, the
  response comes with an `ETag` header. See "ETag Configuration" below for
  details.
* `cacheControl` &mdash; `cache-control` header definition. If present and not
  `false`, every cacheable response comes with the specified header.
* `notFoundPath` &mdash; Optional filesystem path to the file to serve when a
  file/path is not found. The indicated file will get sent back along with a
  `404` ("Not Found") status code.
* `siteDirectory` &mdash; Filesystem directory root for the files to serve.

```js
import { StaticFiles } from '@lactoserv/built-ins';

const applications = [
  {
    name:          'mySite',
    class:         StaticFiles,
    siteDirectory: '/path/to/site',
    notFoundPath:  '/path/to/404.html'
  }
];
```

There are a few notable things which _aren't_ configurable for this application.
These may be configurable in a future version, if there is sufficient and
reasonable demand:

* Content types:
  * The `Content-Type` headers are based on the file extensions of the files
    being served.
  * The mapping from extensions to MIME types is not configurable.
  * Textual file types are always reported to have `charset=utf-8`.
* Caching:
  * The `Last-Modified` response header is always sent.
  * Conditional request headers are honored.
* Ranges:
  * The `Accept-Ranges` response header is always sent.
  * Range request headers are honored, including conditional range requests.
* Directory responses:
  * "Naked" directory paths (i.e. ones that do not end with a slash) are
    redirected to the same path with a final slash appended.
  * Directory paths are responded to with the contents of a file called
    `index.html` in that directory, if it exists. The index file name is not
    configurable.
* These "odd" URL paths all cause not-found responses:
  * Ones with a `..` that would "back out" of the site directory.
  * Ones with an _encoded_ slash in them, that is to say literally `%2F`. (It is
    more trouble than it's worth to try to figure out a way for this to be
    implementable in a non-wacky unambiguous way.)
  * Ones with an internal empty path component, e.g. with `//` somewhere in
    them. Many filesystems will "collapse" multiple slashes away, but we choose
    to err on the side of being conservative and report this as an error rather
    than wade blithely into DWIM territory.
  * End with an empty path component (that is, end with a slash), when the path
    does not in fact correspond to a directory.
* The bodies of error and other non-content responses, other than `404`s, are
  not configurable.
* No files under the `siteDirectory` are filtered out and treated as not found.
  Notably, dotfiles &mdsah; that is, paths where the final component starts with
  a dot (`.`) &mdash; are served when corresponding files are found.

## Built-in Services

**Note about file names:** Several of the services accept a file path as a
configured property, which are typically required to be absolute paths. In
_some_ cases, the final name component of the path is treated as a prefix +
suffix combination, where the prefix is everything before a final dot (`.`), and
the suffix is the final dot and everything that follows. (For example,
`some.file.txt` would be parsed as prefix `some.file` and suffix `.txt`.) These
parsed names are used to construct _actual_ file names by inserting something in
between the prefix and suffix (such as a date stamp and/or a sequence number),
or in some cases just used as-is.

### `MemoryMonitor`

A service which occasionally checks the system's memory usage, and will force a
(hopefully) clean shutdown if memory usage is too high, with an optional grace
period to allow momentary usage spikes. It accepts the following configuration
bindings:

* `checkPeriod` &mdash; How often to check for memory usage being over the
  defined limit, specified as a duration value as described in
  [Durations](#durations). Optional. Minimum `1 sec` (which is frankly way too
  often). Default `5 min` (that is, once every five minutes).
* `gracePeriod` &mdash; Once a memory limit has been reached, how long it is
  allowed to remain at or beyond the maximum before this service takes action,
  specified as a duration value as described in [Durations](#durations). `0` (or
  `null`) to not have a grace period at all. Default `0`. **Note:**: When in the
  middle of a grace period, the service will check memory usage more often than
  `checkPeriod` so as not to miss a significant dip.
* `maxHeapBytes` &mdash; How many bytes of heap is considered "over limit," or
  `null` for no limit on this. The amount counted is `heapTotal + external` from
  `process.memoryUsage()`. Defaults to `null`. **Note:** In order to catch
  probably-unintentional misconfiguration, if a number, must be at least one
  megabyte.
* `maxRssBytes` &mdash; How many bytes of RSS is considered "over limit," or
  `null` for no limit on this. Defaults to `null`. **Note:** In order to catch
  probably-unintentional misconfiguration, if non-`null`, must be at least one
  megabyte.

```js
import { MemoryMonitor } from '@lactoserv/built-ins';

const services = [
  {
    name:         'memory',
    class:        MemoryMonitor,
    checkPeriod:  '5 min',
    gracePeriod:  '1 min',
    maxHeapBytes: 100 * 1024 * 1024,
    maxRssBytes:  150 * 1024 * 1024
  }
];
```

### `ProcessIdFile`

A service which writes a simple text file containing the process ID (number) of
the running system, and which optionally tries to deal with other simultaneous
processes that also write to the same file. The file is written when the system
starts up, when it shuts down (if not killed with extreme prejudice), and
optionally on a periodic basis. It accepts the following configuration bindings:

* `path` &mdash; Path to the file. Must be an absolute path.
* `multiprocess` &mdash; Deal with multiple processes writing to the file?
  Optional and defaults to `false`. If `true`, whenever the file is written, it
  is read first and any process IDs found in it are kept if they are in fact
  still running.
* `updatePeriod` &mdash; How long to wait between each file update, specified as
  a duration value as described in [Durations](#durations), or `null` to
  indicate "never." Optional and defaults to `null`. If specified, the value
  must be at least one second (so as to prevent excessive churn). This value is
  only meaningfully used when `multiprocess` is `true`.

```js
import { ProcessIdFile } from '@lactoserv/built-ins';

const services = [
  {
    name:         'process-id',
    class:        ProcessIdFile,
    path:         '/path/to/var/run/process.txt',
    multiprocess: true,
    updatePeriod: '1 hr'
  }
];
```

### `ProcessInfoFile`

A service which writes out a JSON-format file containing information about the
running system. The file name includes the process ID; and the file is written
when the system starts up, when it shuts down (if not killed with extreme
prejudice), and optionally on a periodic basis. It accepts the following
configuration bindings:

* `path` &mdash; Path to the file, with the final path component modified by
  infixing the process ID.

RELEASES.md
Stable Releases
===============

This is a list of all stable releases that are considered "current." A stable
release is considered current for approximately a year after it is created, or
until an API-compatible release is created that supersedes it, whichever comes
first.

Current releases receive bugfixes and security updates, on an as-needed
best-effort basis. Note that, in the absence of a paid support contract, you
should have no particular expectation about when particular stable releases will
get updated.

**Note:** Version numbers for _stable_ releases follow semantic versioning
principles. Unstable releases do not.

## List of stable releases

* v0.5.20
  * Released 2023-12-15
  * https://github.com/danfuzz/lactoserv/releases/tag/v0.5.20


etc/example-setup/config/config.mjs
// Copyright 2022-2024 the Lactoserv Authors (Dan Bornstein et alia).
// SPDX-License-Identifier: Apache-2.0

import * as fs from 'node:fs/promises';

import { AccessLogToFile, AccessLogToSyslog, EventFan, HostRouter,
  MemoryMonitor, PathRouter, ProcessIdFile, ProcessInfoFile, RateLimiter,
  Redirector, SerialRouter, SimpleResponse, StaticFiles, SyslogToFile }
  from '@lactoserv/webapp-builtins';


const fileUrl  = (path) => new URL(path, import.meta.url);
const filePath = (path) => fileUrl(path).pathname;
const readFile = async (path) => {
  return fs.readFile(fileUrl(path));
}

const VAR_DIR = filePath('../../../out/var');
const LOG_DIR = `${VAR_DIR}/log`;
const RUN_DIR = `${VAR_DIR}/run`;

// Host / certificate bindings.
const hosts = [
  {
    hostnames:   ['localhost', '*.localhost'],
    certificate: await readFile('localhost-cert.pem'),
    privateKey:  await readFile('localhost-key.pem')
  },
  {
    hostnames:   ['*', '127.0.0.1', '::1'],
    selfSigned:  true
  }
];

// Service definitions.
const services = [
  {
    name:         'memory',
    class:        MemoryMonitor,
    checkPeriod:  '10 min',
    gracePeriod:  '1 min',
    maxHeapBytes: 100 * 1024 * 1024,
    maxRssBytes:  150 * 1024 * 1024
  },
  {
    name:         'process',
    class:        ProcessInfoFile,
    path:         `${RUN_DIR}/process.json`,
    updatePeriod: '5 min',
    save: {
      onStart:     true,
      onStop:      true,
      maxOldCount: 10
    }
  },
  {
    name:         'processId',
    class:        ProcessIdFile,
    path:         `${RUN_DIR}/process.txt`,
    multiprocess: true,
    updatePeriod: '5 min'
  },
  {
    name:         'syslog',
    class:        SyslogToFile,
    path:         `${LOG_DIR}/system-log.txt`,
    format:       'human',
    bufferPeriod: '0.25 sec',
    rotate: {
      atSize:      1024 * 1024,
      onStart:     true,
      maxOldBytes: 10 * 1024 * 1024,
      checkPeriod: '1 min'
    }
  },
  {
    name:         'syslogJson',
    class:        SyslogToFile,
    path:         `${LOG_DIR}/system-log.json`,
    format:       'json',
    bufferPeriod: '0.25 sec',
    rotate: {
      atSize:      2 * 1024 * 1024,
      onStart:     true,
      onReload:    true,
      onStop:      true,
      maxOldBytes: 10 * 1024 * 1024,
      maxOldCount: 10,
      checkPeriod: '1 min'
    }
  },
  {
    name:  'accessFile',
    class: AccessLogToFile,
    path:  `${LOG_DIR}/access-log.txt`,
    bufferPeriod: '0.25 sec',
    rotate: {
      atSize:       10000,
      maxOldCount:  10,
      checkPeriod:  '1 min'
    }
  },
  {
    name:  'accessSyslog',
    class: AccessLogToSyslog
  },
  {
    name:     'accessLog',
    class:    EventFan,
    services: ['accessFile', 'accessSyslog']
  },
  {
    name:        'limiter',
    class:       RateLimiter,
    connections: {
      maxBurstSize: 10,
      flowRate:     '3 per sec',
      maxQueueSize: 25
    },
    requests: {
      maxBurstSize: 20,
      flowRate:     '600 per min',
      maxQueueSize: 100
    },
    data: {
      maxBurstSize:      1024 * 1024,           // 1MB.
      flowRate:          `${100 * 1024} / sec`, // 100kB.
      maxQueueGrantSize: 50 * 1024,             // 50kB.
      maxQueueSize:      2 * 1024 * 1024        // 2MB.
    }
  }
];

// Application definitions.
const applications = [
  // Main apps.

  {
    name:         'myWackyRedirector',
    class:        Redirector,
    statusCode:   308,
    target:       'https://localhost:8443/resp/',
    cacheControl: { public: true, maxAge: '5 min' }
  },

  {
    name:  'mySite',
    class: HostRouter,
    hosts: {
      '*':         'myPaths',
      '127.0.0.1': 'mySeries'
    }
  },
  {
    name:  'myPaths',
    class: PathRouter,
    paths: {
      '/*':                 'myStaticFun',
      '/bonk/*':            'myStaticFun',
      '/florp/*':           'myStaticFunNo404',
      '/resp/empty-body/*': 'responseEmptyBody',
      '/resp/no-body/*':    'responseNoBody',
      '/resp/dir-only/':    'responseDirOnly',
      '/resp/one':          'responseOne',
      '/resp/two':          'responseTwo'
    }
  },
  {
    name:  'mySeries',
    class: SerialRouter,
    applications: [
      'myStaticFunNo404',
      'responseNotFound'
    ]
  },

  // Component apps used by the above.

  {
    name:           'myStaticFun',
    class:          StaticFiles,
    siteDirectory:  filePath('../site'),
    notFoundPath:   filePath('../site-extra/not-found.html'),
    cacheControl:   { public: true, maxAge: '5 min' },
    etag:           { dataOnly: true, hashLength: 20 },
    maxQueryLength: 20
  },
  {
    name:          'myStaticFunNo404',
    class:         StaticFiles,
    siteDirectory: filePath('../site'),
    cacheControl:  { public: true, maxAge: '5 min' },
    etag:          { dataOnly: true, hashLength: 20 }
  },
  {
    name:                'responseEmptyBody',
    class:               SimpleResponse,
    filePath:            filePath('../site-extra/empty-file.txt'),
    cacheControl:        'public, immutable, max-age=600',
    maxPathLength:       2,
    redirectDirectories: true
  },
  {
    name:        'responseNotFound',
    class:       SimpleResponse,
    etag:        false,
    body :       'Sorry! Not found!\n',
    contentType: 'text/plain',
    statusCode:  404
  },
  {
    name:          'responseNoBody',
    class:         SimpleResponse,
    cacheControl:  { public: true, immutable: true, maxAge: '11 min' },
    etag:          true,
    maxPathLength: 2,
  },
  {
    name:         'responseDirOnly',
    class:        SimpleResponse,
    contentType:  'text/plain',
    body:         'I am a directory!\n',
    cacheControl: { public: true, immutable: true, maxAge: '12 min'  },
    etag:         true
  },
  {
    name:         'responseOne',
    class:        SimpleResponse,
    contentType:  'text/plain',
    body:         'One!\n',
    cacheControl: { public: true, immutable: true, maxAge: '12 min'  },
    etag: {
      hashAlgorithm: 'sha1',
      hashLength:    12,
      tagForm:       'weak'
    }
  },
  {
    name:                'responseTwo',
    class:               SimpleResponse,
    contentType:         'text/html',
    body:                '<html><body><h1>Two!</h1></body></html>\n',
    cacheControl:        { public: true, immutable: true, maxAge: '13 min'  },
    etag:                true
  }
];

// Endpoint defintions, including mount points for applications.
const endpoints = [
  {
    name:      'insecure',
    protocol:  'http',
    interface: '*:8080',
    services: {
      accessLog:   'accessLog',
      rateLimiter: 'limiter'
    },
    application: 'myWackyRedirector'
  },
  {
    name:      'secure',
    protocol:  'http2',
    hostnames: ['*'],
    interface: '*:8443',
    services: {
      accessLog:   'accessLog',
      rateLimiter: 'limiter'
    },
    application: 'mySite'
  },
  {
    name: 'alsoSecure',
    protocol:  'https',
    hostnames: ['*'],
    interface: '*:8444',
    services: {
      accessLog: 'accessLog'
    },
    application: 'mySeries'
  }
];

const config = {
  applications,
  endpoints,
  hosts,
  services
};

export default config;
